File: Java/13_Stack & Queue/E232. Implement Queue using Stacks.java
// 232. Implement Queue using Stacks (0(1) - amortized method)
// https://leetcode.com/problems/implement-queue-using-stacks/
// Implement a first in first out (FIFO) queue using only two stacks.
// The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).
// Implement the MyQueue class:
// 1. Push element x to the back of queue.
// 2. Remove the element from the front of queue and returns it.
// 3. Get the front element.
// 4. Returns true if the queue is empty, false otherwise.

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */
import java.util.Stack;

class MyQueue {

    Stack<Integer> stack1;
    Stack<Integer> stack2;

    public MyQueue() {
        stack1 = new Stack<>();
        stack2 = new Stack<>();
    }

    public void push(int x) {
        stack1.push(x);
    }

    public int pop() {
        if (stack2.isEmpty()) {
            while (!stack1.isEmpty()) {
                stack2.push(stack1.pop());
            }
        }
        return stack2.pop();
    }

    public int peek() {
        if (stack2.isEmpty()) {
            while (!stack1.isEmpty()) {
                stack2.push(stack1.pop());
            }
        }
        return stack2.peek();
    }

    public boolean empty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }
    
    public static void main(String[] args) {
        MyQueue queue = new MyQueue();
        queue.push(1);
        queue.push(2);
        System.out.println(queue.peek()); // Output: 1
        System.out.println(queue.pop()); // Output: 1
        System.out.println(queue.empty()); // Output: false
    }
}

File: JavaScript/13_Stack & Queue/E232. Implement Queue using Stacks.js
// 232. Implement Queue using Stacks
// https://leetcode.com/problems/implement-queue-using-stacks/

/**
 * Creates a queue using stacks (functional approach)
 * Uses two stacks - amortized O(1) for all operations
 * @returns {Object} Queue operations
 */
function createMyQueue() {
    let stackIn = [];
    let stackOut = [];

    function push(x) {
        stackIn.push(x);
    }

    function pop() {
        peek(); // Ensure stackOut has elements
        return stackOut.pop();
    }

    function peek() {
        if (stackOut.length === 0) {
            while (stackIn.length > 0) {
                stackOut.push(stackIn.pop());
            }
        }
        return stackOut[stackOut.length - 1];
    }

    function empty() {
        return stackIn.length === 0 && stackOut.length === 0;
    }

    return { push, pop, peek, empty };
}

// Test cases
const myQueue = createMyQueue();
myQueue.push(1);
myQueue.push(2);
console.log(myQueue.peek());  // 1
console.log(myQueue.pop());   // 1
console.log(myQueue.empty()); // false
myQueue.push(3);
console.log(myQueue.pop());   // 2
console.log(myQueue.pop());   // 3
console.log(myQueue.empty()); // true
